# **演習：関数とヘルプの表示**

**This notebook is an exercise in the [Python](https://www.kaggle.com/learn/python) course.  You can reference the tutorial at [this link](https://www.kaggle.com/colinmorris/functions-and-getting-help).**

---


**このノートは、[Python](https://www.kaggle.com/learn/python) コースの練習問題です。チュートリアルは、[このリンク](https://www.kaggle.com/colinmorris/functions-and-getting-help)**で参照できます。

---


Functions are powerful. Try writing some yourself.

As before, don't forget to run the setup code below before jumping into question 1.

関数は強力です。自分で書いてみてください。

前と同様に、質問1に進む前に、以下のセットアップコードを実行することを忘れないでください。

In [1]:
# SETUP.現時点では、このコードが何をするのか、どのように動作するのかについて心配する必要はありません。
from learntools.core import binder; binder.bind(globals())
from learntools.python.ex2 import *
print('Setup complete.')

Setup complete.


# 1.

Complete the body of the following function according to its docstring.

HINT: Python has a built-in function `round`.

# 1.

docstringに従って、次の関数の本体を完成させます。

ヒント:Pythonには組み込み関数`round`があります。

In [2]:
def round_to_two_places(num):
    """指定された数値を小数点以下2桁に丸めて返します。
    
    >>> round_to_two_places(3.14159)
    3.14
    """
    return round(num,2)
    
    # この本文を独自のコードに置き換えます。
    #  ("pass"は文字通り何もしないキーワードです。プレースホルダーとして使用しました。
    # コードブロックを開始した後、Pythonは少なくとも1行のコードを必要とするため)
    # pass

# 答えをチェック
q1.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [3]:
# Uncomment the following for a hint
q1.hint()
# Or uncomment the following to peek at the solution
q1.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> Run `help(round)` in the console (or in a code cell) to learn more about the round function. You'll need to use the function's optional second argument.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
return round(num, 2)
```

# 2.
The help for `round` says that `ndigits` (the second argument) may be negative.
What do you think will happen when it is? Try some examples in the following cell.

# 2.
`round`のヘルプによると、`ndigits` (2番目の引数) は負の値になる可能性があります。
そうなったらどうなると思いますか?次のセルにいくつかの例を示します。

In [4]:
# テストコードをここに入力します。
print(
    round(123.145,-1),
    round(123.145,-2),
    round(456.145,-1),
    round(456.145,-2),
)

120.0 100.0 460.0 500.0


Can you think of a case where this would be useful?  Once you're ready, run the code cell below to see the answer and to receive credit for completing the problem.

これが役に立つケースを考えられますか?準備ができたら、下のコードセルを実行して答えを確認し、問題を完了したことを示すクレジットを受け取ります。

In [5]:
# 答えをチェックしてください(クレジットを受け取るには、このコードセルを実行してください。)。
q2.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> As you've seen, `ndigits=-1` rounds to the nearest 10, `ndigits=-2` rounds to the nearest 100 and so on. Where might this be useful? Suppose we're dealing with large numbers:

> The area of Finland is 338,424 km²  
> The area of Greenland is 2,166,086 km²

We probably don't care whether it's really 338,424, or 338,425, or 338,177. All those digits of accuracy are just distracting. We can chop them off by calling `round()` with `ndigits=-3`:

> The area of Finland is 338,000 km²  
> The area of Greenland is 2,166,000 km²

(We'll talk about how we would get the commas later when we talk about string formatting :))


---
解決方法:これまで見てきたように、ndigits=-1を10の近似値に、ndigits=-2を100の近似値に丸めます。これはどのような場合に役立つでしょうか。大きな数を扱っているとします。

> フィンランドの面積は338,424平方キロメートルです。  
> グリーンランドの面積は2,166,086平方キロメートルです。

実際に338,424年だろうが338,425年だろうが338,177年だろうがおそらく気にしません。精度の桁数が気になるだけです。ndigits=-3を指定してround () を呼び出すことで、これらを取り除くことができます。

> フィンランドの面積は338,000平方キロメートルです。  
> グリーンランドの面積は2,166,000平方キロメートルです。

(カンマを取得する方法については、後で文字列の書式設定について説明するときに説明します。)

# 3.

In the previous exercise, the candy-sharing friends Alice, Bob and Carol tried to split candies evenly. For the sake of their friendship, any candies left over would be smashed. For example, if they collectively bring home 91 candies, they'll take 30 each and smash 1.

Below is a simple function that will calculate the number of candies to smash for *any* number of total candies.

Modify it so that it optionally takes a second argument representing the number of friends the candies are being split between. If no second argument is provided, it should assume 3 friends, as before.

Update the docstring to reflect this new behaviour.

# 3.

前の練習では、キャンディを分け合う友達のAlice、Bob、Carolがキャンディを均等に分けようとしました。彼らの友情のために、残ったキャンディーは砕かれるだろう。例えば、みんなで91個の飴を持って帰ったら、30個ずつ持って行って1個潰します。

以下の単純な関数は、「任意の」個数のキャンディを潰す個数を計算します。

必要に応じて、キャンディを分割する友人の数を表す2番目の引数を取るように変更します。2番目の引数が指定されていない場合は、以前と同様に3つのフレンドを想定する必要があります。

この新しい動作を反映するようにdocstringを更新します。

In [6]:
def to_smash(total_candies, num=3):
    """配布後につぶさなければならない残りのキャンディの数を返します。
    num人の友人の間で均等に与えられた数のキャンディー。
    
    >>> to_smash(91,3)
    1
    """
    
    
    return total_candies % num

# Check your answer
q3.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [7]:
q3.hint()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> Refer to the section of the last tutorial notebook where we talked about default arguments

In [8]:
q3.solution()

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
def to_smash(total_candies, n_friends=3):
    return total_candies % n_friends
```

# 4. (Optional)

It may not be fun, but reading and understanding error messages will be an important part of your Python career.

Each code cell below contains some commented buggy code. For each cell...

1. Read the code and predict what you think will happen when it's run.
2. Then uncomment the code and run it to see what happens. (**Tip**: In the kernel editor, you can highlight several lines and press `ctrl`+`/` to toggle commenting.)
3. Fix the code (so that it accomplishes its intended purpose without throwing an exception)

<!-- TODO: should this be autochecked? Delta is probably pretty small. -->

# 4 (任意)

楽しくないかもしれませんが、エラーメッセージを読んで理解することは、Pythonキャリアの重要な部分になるでしょう。

以下の各コードセルには、コメント付きのバグのあるコードが含まれています。各セルに対して...

1. コードを読んで、実行時に何が起こるかを予測します。
2. 次に、コードのコメントを解除して実行し、何が起こるかを確認します。(**ヒント**:カーネルエディタで、いくつかの行を強調表示して`ctrl`+`/`を押すと、コメントを切り替えることができます。)
3. コードを修正しました (例外をスローせずに意図した目的を達成できるように) 。

<!--  TODO:これを自動チェックしますか?デルタはかなり小さいでしょう。  -->

In [9]:
round_to_two_places(9.9999)

10.0

In [10]:
x = -10
y = 5
# 上の2つの変数のうち、絶対値が最も小さいのはどれですか。
smallest_abs = min(abs(x),abs(y))
smallest_abs

5

In [11]:
def f(x):
    y = abs(x)
    return y

print(f(5))

5


# Keep Going

Nice job with the code. Next up, you'll learn about *conditionals*, which you'll need to **[write interesting programs](https://www.kaggle.com/colinmorris/booleans-and-conditionals)**. 

# 続けて

コードはよくやった。次に、**[面白いプログラムを書く](https://www.kaggle.com/colinmorris/booleans-and-conditionals)**。

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/python/discussion) to chat with other learners.*

---




*ご質問やご意見がありますか?[Course Discussion Forum](https://www.kaggle.com/learn/python/discussion)にアクセスして、他の学習者とチャットしてください。*